### DISCUSSION TITLE
Can we install slinger on oracle cloud vm?

### DISCUSSION THREAD

**Discussion by f520k510** [GerryDazoo/Slinger#49] [Ideas]
I was thinking about getting a mini pc or rasberri pi to run linux for the purpose of running slinger server.  Then I think of the idea if we can run it on oracle cloud server, which has free 2 vm.  I tried hard from last night to this morning but couldn't get it working.

I wonder if it is too much to ask anyway.  Having getting it running on mac, and windows with python and exe.

I really appreciate the great work and the effort spent on coding this piece of software.

[slingbox@instance-20220911-0539 ~]$ python3 slingbox_server.py
Version : 3.07 Running on Linux-5.15.0-1.43.4.1.el9uek.x86_64-x86_64-with-glibc2.34 pid= 47310
Using config file config.ini
Connection Manager Running on port 8080 with 10 max streams....
Streamer Running:  10 config.ini 8080
starting up on  port 8080
Checking for slingbox at ('125.228.xx.xx', 5216)
Reading Custom Remote definition from remote.txt
Error reading remote definition file, using defaults
 * Serving Flask app 'slingbox_server'
 * Debug mode: off
WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:8081
 * Running on http://10.0.0.254:8081
Press CTRL+C to quit
09/11/2022, 01:44:43  Registering Slingboxes {'xxxxxxxxxxxxxxx7CD334': 8080}
Getting external IP address
Error connecting to  ('125.228.xx.xx', 5216)
No valid slingbox ip info found in config.ini
Finding Slingbox on local network. My IP Info =  02:00:17:xx:xx:xx
Finding Slingbox on local network. My IP Info =  10.0.0.254
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib64/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "/home/slingbox/slingbox_server.py", line 539, in streamer
    for port in stati.keys:
TypeError: 'builtin_function_or_method' object is not iterable
Register: Error Getting Exteral IP Address, Retrying in 10 minutes

Please look into it only if you have time.  I got the mac version running now.  Thanks.

**Comment by GerryDazoo** [comment]
I believe you've uncovered a bug. I'll look into it. For now comment out your FInderid in the config.ini to get past this.
In the meantime it looks like the slingbox isn't reachable (Error connecting to ('125.228.xx.xx', 5216).
It looks like you might have firewall issues, that also explains this error Register: Error Getting Exteral IP Address, Retrying in 10 minutes,

**Comment by f520k510** [comment]
Hi,
I just wanted to report that I have successfully installed slinger on a oracle cloud vm.
I followed everything that was in the readme.txt.
I also use the following commands to open all ports:
sudo iptables -P INPUT ACCEPT
sudo iptables -P OUTPUT ACCEPT
sudo iptables -P FORWARD ACCEPT
sudo iptables -F 

I know it is not a good practice but it works.

I think it is a very ideal solution:  free, fast, and run 24hrs/7days.

**Reply by flipper2222222** [reply]
f520k510 - this sounds interesting , could you please list in point form some more details on how you achieved this free solution? thanks in advance

**Reply by f520k510** [reply]
Okay.
1.  apply for an oracle cloud (OCI) account.  You need a credit card, but they won't charge you until you subscribe to their paid service.
2.  create a vm instance.  I chose centos 8.  save the private key.
3.    scp -i private_key -r /path to slinger/ opc@your_vm_wan_ip:/home/opc/
4.  login from terminal (or putty for windows) with ssh -i private_key opc@your_vm_wan_ip  (you can check its ip after creating the vm instance.
5  nano config.ini and edit the necessary fields.  ipaddress=slingbox wan ip port=slingbox port finderid is needed, too.
6  issue the following commands on the ssh:
    sudo iptables -P INPUT ACCEPT
    sudo iptables -P OUTPUT ACCEPT
    sudo iptables -P FORWARD ACCEPT
    sudo iptables -F
7  install python and modules flask, netifaces.
8 at ssh:  python3 slingbox_server.py
9  if installed correctly, it will say "register successful"
10 try view with vlc at http://your vm's wan ip:8080/slingbox
11  nano sling.service and edit the line ExecStart=/usr/bin/python -u slingbox_server.py to ExecStart=/usr/bin/python3 -u slingbox_server.py
12  at ssh:  sudo adduser slingbox
                  sudo passwd slingbox    (give it a password)
13  at ssh:  sudo cp sling.service /etc/systemd/system/.
                   sudo cp sling.service /home/slingbox/.
                   sudo cp slingbox_server.py /home/slingbox/.
                   sudo cp config.ini /home/slingbox/.
                   sudo cp remote.txt /home/slingbox/.
14  sudo systemctl daemon-reload
      sudo systemctl start sling.service
      sudo systemctl enable sling.service
15  sudo yum install iptables-services
      sudo systemctl start iptables
      sudo systemctl enable iptables

These are just dumped from my head.  I hope it works for you.

**Reply by GerryDazoo** [reply]
A few comments. The finderid is only useful if you don't have a static IP address, which is not the case here. Configuring this needlessly put additional load on my little AWS instance for no reason.

Thanks for the step by step. You really should only open the needed ports in your IPtables firewall rules. Your VM is going to be port scanned continuously.  I think I've got some sample rules laying around here somewhere, I'll post them.

Also, from my experience with AWS, I noticed that a lot of the port scanners only check lower port number and other well know ports. You may have noticed from my other posts I've been using port 65432. This has reduced the number of invalid streaming requests by a order of magnitude. They usually give up before getting to these high port numbers.

**Reply by htwonpoke** [reply]
I have been trying to replicate your oracle setup but I am running into some issues. My output looks like:

Version : 3.07 Running on Linux-4.18.0-408.el8.x86_64-x86_64-with-centos-8 pid= 2262
Using config file config.ini
Connection Manager Running on port 65432 with 10 max streams....
Streamer Running:  10 config.ini 65432
starting up on  port 65432
Checking for slingbox at ('x.x.x.x', 5201)
('x.x.x.x', 5201) OK
Using slingbox at  ('x.x.x.x', 5201)
Streamer:   Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
 * Serving Flask app 'slingbox_server' (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on all addresses.
   WARNING: This is a development server. Do not use it in a production deployment.
 * Running on http://x.x.x.x:65433/ (Press CTRL+C to quit)
09/24/2022, 04:06:24  Registering Slingboxes {'946D6E4D...': 65432}
Getting external IP address
Redirector, my external IP address X.X.X.X
PING failure ('X.X.X.X, 65432) [Errno 110] Connection timed out
No working external portmaps. Not registering slingboxes
Retrying in 10 minutes

I copied my files directly over from my working setup on my mac. Any ideas of why I keep getting this? I think I copied all of your iptables stuff correctly.

**Reply by f520k510** [reply]
I am not sure.  Try commenting out finderid in config.ini.  It's not needed as pointed out by Gerry.

**Reply by GerryDazoo** [reply]
yes. f520k510. Comment out the finderid. You only need it if you don't have a static IP address. But that log won't stop the server from streaming.

**Reply by htwonpoke** [reply]
Sorry, I go back ad forth between commenting that Finder ID out. For some reason on my mac setup if I have to have this Finder ID to work. Which I thought my IP would be static since im accessing the slingbox remotely. 

But when I comment out the finder id with the oracle setup I get:

Version : 3.07 Running on Linux-4.18.0-408.el8.x86_64-x86_64-with-centos-8 pid= 2885
Using config file config.ini
Connection Manager Running on port 8080 with 10 max streams....
Streamer Running:  10 config.ini 8080
Checking for slingbox at ('X.X.X.X', 5201)
starting up on  port 8080
('X.X.X.X', 5201) OK
Using slingbox at  ('X.X.X.X', 5201)
Streamer:   Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
 * Serving Flask app 'slingbox_server' (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on all addresses.
   WARNING: This is a development server. Do not use it in a production deployment.
 * Running on http://10.0.0.58:8081/ (Press CTRL+C to quit)

It seems to hang at this point. I did change it to just use the default 8080 from the config file.


 Could it be that I didnt set up my iptables correctly? My Tables look like:

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         
ACCEPT     all  --  anywhere             anywhere             state RELATED,ESTABLISHED
ACCEPT     icmp --  anywhere             anywhere            
ACCEPT     all  --  anywhere             anywhere            
ACCEPT     tcp  --  anywhere             anywhere             state NEW tcp dpt:ssh
REJECT     all  --  anywhere             anywhere             reject-with icmp-host-prohibited

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         
REJECT     all  --  anywhere             anywhere             reject-with icmp-host-prohibited

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
 

Does this look ok? Thanks a lot for the help.

**Reply by htwonpoke** [reply]
Just to update, I figured it out.I had to update the ingress rules for oracle instance and then everything started working. Thanks again for getting me started!

**Comment by GerryDazoo** [comment]
Here's some iptables rules from my AWS instance.   65432 is for my server. 22 is ssh. I think this is all you need

-A INPUT -m state --state RELATED,ESTABLISHED -j ACCEPT
-A INPUT -i lo -j ACCEPT
-A INPUT -p tcp -m state --state NEW -m tcp --dport 22 -j ACCEPT
-A INPUT -p tcp -m state --state NEW -m tcp --dport 65432 -j ACCEPT

**Reply by f520k510** [reply]
Thanks.  I just commented out the finder id and updated iptables rules.

**Reply by GerryDazoo** [reply]
Actually you'll need these two rules at the end
-A INPUT -j REJECT --reject-with icmp-host-prohibited
-A FORWARD -j REJECT --reject-with icmp-host-prohibited
